rubix-spi/src/test/java/com/qubole/rubix/spi/client/TestPoolingClient.java

```diff
@@ -0,0 +1,173 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License. See accompanying LICENSE file.
+ */
+package com.qubole.rubix.spi.client;
+
+import com.qubole.rubix.spi.CacheConfig;
+import com.qubole.rubix.spi.fop.ObjectPool;
+import com.qubole.rubix.spi.fop.SocketObjectFactory;
+import com.qubole.rubix.spi.thrift.Request;
+import com.qubole.rubix.spi.thrift.TestingService;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.thrift.TException;
+import org.apache.thrift.server.TServer;
+import org.apache.thrift.server.TThreadPoolServer;
+import org.apache.thrift.transport.TServerSocket;
+import org.apache.thrift.transport.TServerTransport;
+import org.apache.thrift.transport.TTransportException;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.qubole.rubix.spi.CacheConfig.getServerMaxThreads;
+import static org.testng.Assert.assertTrue;
+
+public class TestPoolingClient
+{
+  private static final int PORT = 8656;
+  private static TServer server;
+  private static Configuration conf;
+
+  @BeforeClass
+  public static void setup()
+  {
+    conf = new Configuration();
+    // min=max=1 => we get same transport always
+    CacheConfig.setTranportPoolMinSize(conf, 1);
+    CacheConfig.setTranportPoolMaxSize(conf, 1);
+    CacheConfig.setServerSocketTimeout(conf, 500);
+    CacheConfig.setServerConnectTimeout(conf, 1000);
+
+    startServerAsync(conf);
+  }
+
+  @Test
+  public void testRetryHappens()
+      throws TException, InterruptedException
+  {
+    ObjectPool pool = SocketObjectFactory.createSocketObjectPool(conf, "localhost", PORT);
+    RetryingPoolingTestClient client = new RetryingPoolingTestClient(
+        20,
+        conf,
+        "localhost",
+        pool.borrowObject("localhost", conf));
+
+    //  issue a request which is sure to hit SocketTimeout exception
+    final Request request = new Request("_Message_", 2);
+    final AtomicReference<String> message = new AtomicReference<>();
+
+    // Start a Thread that removes sleep after some time
+    // If retries are in place, the response will be received after request has been modified
+    new Thread(new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        try {
+          Thread.sleep(2000);
+        }
+        catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+        assertTrue(message.get() == null, "Request should not have finished");
+        request.setSleepSecs(0);
+      }
+    }).start();
+
+    message.set(client.echo(request));
+
+    // given enough retries, we should have the response
```

Comment:
shouldn't we wait for the thread to finish?

Reply:
with the number of retries=20 and socketTimeout=500msec, the thread is going to finish before retries finish. In case it doesnt, the test will fail anyways so we dont need to wait.